lib/vector-core/src/event/metadata.rs

```diff
@@ -345,6 +347,33 @@ impl EventMetadata {
     pub fn set_schema_definition(&mut self, definition: &Arc<schema::Definition>) {
         self.schema_definition = Arc::clone(definition);
     }
+
+    /// Helper function to add a semantic meaning to the schema definition.
+    ///
+    /// This replaces the common code sequence of:
+    /// ```
+    /// let new_schema = log_event
+    ///     .metadata()
+    ///     .schema_definition()
+    ///     .as_ref()
+    ///     .clone()
+    ///     .with_meaning(target_path, meaning);
+    /// log_event
+    ///     .metadata_mut()
+    ///     .set_schema_definition(new_schema);
+    /// ````
+    ///
+    /// Note: This plays games with uninitialized memory to avoid cloning a new schema if possible.
+    pub fn add_schema_meaning(&mut self, target_path: OwnedTargetPath, meaning: &str) {
+        // This use of uninitialized invalid values is safe because we immediately replace the value
+        // with a modified schema. We could probably avoid this `unsafe` by owning `self`, but that
+        // makes the calling convention harder here, since most callers will only have a `&mut self`.
+        #[allow(invalid_value, clippy::uninit_assumed_init)]
+        let dummy = unsafe { mem::MaybeUninit::uninit().assume_init() };
```

Comment:
I'm not actually sure this is OK.

It's OK in the sense that we don't actually access `self.schema_definition` until we end up replacing it with the newly-modified version. We're also dropping the dummy version that holds its place as we do that, though, and that dummy version is potentially _not_ OK to drop in an uninitialized state.

Could we just use `Arc::make_mut` instead and add an accessor for setting the meaning instead of using the fluent builder-style `with_meaning`? That would accomplish the same thing of cloning it if it was shared but we could skip the replace stuff going on here.

Reply:
Yeah, that's so much simpler, and there's even a `fn add_meaning` already that `with_meaning` calls, it just wasn't `pub`. I'm not sure why I didn't just make that public instead.